src/software/ai/hl/stp/stp.cpp

```diff
@@ -219,19 +219,52 @@ std::map<std::shared_ptr<const Tactic>, Robot> STP::assignRobotsToTactics(
     ConstPriorityTacticVector tactics, const World& world,
     bool automatically_assign_goalie)
 {
-    robot_tactic_assignment.clear();
+    std::map<std::shared_ptr<const Tactic>, Robot> new_tactic_assignment =
+        std::map<std::shared_ptr<const Tactic>, Robot>();
 
     std::optional<Robot> goalie_robot = world.friendlyTeam().goalie();
     std::vector<Robot> robots         = world.friendlyTeam().getAllRobots();
 
+    double old_assignment              = 0;
+    unsigned int new_tactics           = 0;
+    bool recalculate_tactic_assignment = false;
+    for (auto tactic_vector : tactics)
+    {
+        for (auto tactic : tactic_vector)
+        {
+            if (robot_tactic_assignment.find(tactic) != robot_tactic_assignment.end())
+            {
+                old_assignment +=
+                    tactic->calculateRobotCost(robot_tactic_assignment.at(tactic), world);
+            }
+            else
+            {
+                recalculate_tactic_assignment = true;
+            }
+            ++new_tactics;
+        }
+    }
+
     if (goalie_robot && automatically_assign_goalie)
     {
-        robot_tactic_assignment.emplace(goalie_tactic, goalie_robot.value());
+        ++new_tactics;
+        if (robot_tactic_assignment.find(goalie_tactic) == robot_tactic_assignment.end())
+        {
+            recalculate_tactic_assignment = true;
+        }
+        new_tactic_assignment.emplace(goalie_tactic, goalie_robot.value());
```

Comment:
Looking for suggestions to make this code clearer.
Basically, force re-calculate the tactic assignment for this tick if:
- requested tactic vector size is different from previous tactic vector size
- requested tactic wasn't assigned previously